browser-test/src/reporters/file_placement_reporter.ts

```diff
@@ -0,0 +1,158 @@
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
+/* eslint-disable @typescript-eslint/no-unsafe-call */
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import fs = require('fs')
+import path = require('path')
+import sharp = require('sharp')
+import type {Reporter} from '@playwright/test/reporter'
+
+/**
+ * A custom playwright reporter that runs after all reporters have ended their processing. It
+ * handles moving files to the `diff_output` and `updated_snapshots` folders.
+ */
+class FilePlacementReporter implements Reporter {
+  /**
+   * Called immediately before test runner exists. At this point all the reporters have received the reporter.onEnd()
+   * signal, so all the reports should be build. You can run the code that uploads the reports in this hook.
+   * See https://playwright.dev/docs/api/class-reporter#reporter-on-exit
+   *
+   * @override
+   */
+  async onExit() {
+    await this.populateImages()
+  }
+
+  /**
+   * Read the json report and kickstart the process of handling the output
+   */
+  async populateImages() {
+    const jsonObj = JSON.parse(
+      fs.readFileSync('./tmp/json-output/results.json', 'utf8'),
+    )
+    await this.walkSuites(jsonObj['suites'])
+  }
+
+  /**
+   * Recurse through nested suites arrays to get to the specs
+   * @param suites An array of elements from a suites property from the playwright json report output
+   */
+  async walkSuites(suites: any) {
+    for (const suite of suites) {
+      if (suite.suites !== undefined) {
+        await this.walkSuites(suite.suites)
+      }
+
+      await this.processSpecs(suite.specs)
+    }
+  }
+
+  /**
+   * Iterate over specs and for ones from failed tests place updated and diff images in desired folders
+   * @param specs An array of elements from a specs property from the playwright json report output
+   */
+  async processSpecs(specs: any) {
+    for (const spec of specs) {
+      if (spec.ok === true) {
+        continue
+      }
+
+      for (const test of spec.tests) {
+        for (const result of test.results) {
+          const expected = result.attachments[0]
+          const actual = result.attachments[1]
+          const diff = result.attachments[2]
+
+          this.copyActualImageToUpdatedSnapshotsFolder(actual)
+          await this.createCompositeDiffAndCopyToDiffOutputFolder(
+            expected,
+            actual,
+            diff,
+          )
+        }
+      }
+    }
+  }
+
+  /**
+   * Copy the image that was made by the tests to the updated_snapshots folder
+   * @param actual The {@link Attachment} info for the actual image that was generated by the test
+   */
+  copyActualImageToUpdatedSnapshotsFolder(actual: Attachment) {
+    // Copy actual file to updated_snapshots
+    const updatedSnapshotFile = path.join(
+      './updated_snapshots',
+      actual.name.replace('-actual.png', '-received.png'),
```

Comment:
question (non-blocking): Why rename this from `-actual` to `-received`?

Reply:
It maintains compatibility with the naming from before when we still used Jest. We need to update other scripts that look for `-received`. A step I hadn't gotten around to doing yet. I added a comment.